src/s3_plugin.js

```diff
@@ -312,4 +322,92 @@ module.exports = class S3Plugin {
       }
     })
   }
+
+  setIndex() {
```

Comment:
This function is a bit long, can you split it out into some helpers? There's a lot of nested if statements too.